sdk/schemaregistry/azure-data-schemaregistry-jsonschema/src/main/java/com/azure/data/schemaregistry/jsonschema/SchemaRegistryJsonSchemaSerializer.java

```diff
@@ -0,0 +1,345 @@
+     */
+    public <T extends MessageContent> Mono<T> serializeAsync(Object object,
+        TypeReference<T> typeReference, Function<BinaryData, T> messageFactory) {
+
+        if (object == null) {
+            return monoError(logger, new NullPointerException(
+                "Null object, behavior should be defined in concrete serializer implementation."));
+        } else if (typeReference == null) {
+            return monoError(logger, new NullPointerException("'typeReference' cannot be null."));
+        }
+
+        final String schemaFullName = object.getClass().getName();
+        final Optional<Constructor<?>> constructor =
+            Arrays.stream(typeReference.getJavaClass().getDeclaredConstructors())
+                .filter(c -> c.getParameterCount() == 0)
+                .findFirst();
+
+        if (!constructor.isPresent() && messageFactory == null) {
+            return Mono.error(new IllegalArgumentException(typeReference.getJavaClass() + "does not have have a no-arg "
+                + "constructor to create a new instance of T with. Use the overload that accepts 'messageFactory'."));
+        }
+
+        final Function<BinaryData, T> messageFactoryToUse = messageFactory != null ? messageFactory
+            : binaryData -> {
+            final T instance = createNoArgumentInstance(typeReference);
+            instance.setBodyAsBinaryData(binaryData);
+
+            return instance;
+        };
+
+        final byte[] encoded;
+        try {
+            encoded = serializer.serializeToBytes(object, ENCODING);
+        } catch (IOException e) {
+            return Mono.error(new UncheckedIOException("An error occurred while attempting to serialize the data.", e));
+        }
+
+        final T serializedMessage = messageFactoryToUse.apply(BinaryData.fromBytes(encoded));
+
+        String schemaDefinition;
+        try {
+            schemaDefinition = schemaGenerator.generateSchema(TypeReference.createInstance(object.getClass()));
+        } catch (Exception exception) {
+            logger.atError()
+                    .addKeyValue("type", schemaFullName)
+                    .log(() -> "An error occurred while attempting to generate the schema.", exception);
+            return Mono.error(exception);
+        }
+
+        if (schemaDefinition == null) {
+            logger.atWarning().addKeyValue("type", schemaFullName)
+                    .log("Schema returned from generator was null.");
+            return Mono.error(new IllegalArgumentException("JSON schema cannot be null. Type: " + schemaFullName));
+        }
+
+        return this.schemaCache.getSchemaId(schemaFullName, schemaDefinition)
+            .handle((schemaId, sink) -> {
+                try {
+                    serializedMessage.setContentType(CONTENT_TYPE + "+" + schemaId);
+
+                    sink.next(serializedMessage);
+                    sink.complete();
```

Comment:
nit; because it's Mono, the next() call should be sufficient to successfully _complete_

Reply:
Good catch.